drivers/lcd/apa102.c

```diff
@@ -0,0 +1,667 @@
+   */
+
+  struct apa102_ledstrip_s fb[APA102_FBSIZE];
+};
+
+/****************************************************************************
+ * Private Function Protototypes
+ ****************************************************************************/
+
+/* LCD Data Transfer Methods */
+
+static int apa102_putrun(FAR struct lcd_dev_s *dev, fb_coord_t row,
+                         fb_coord_t col, FAR const uint8_t *buffer,
+                         size_t npixels);
+static int apa102_getrun(FAR struct lcd_dev_s *dev, fb_coord_t row,
+                         fb_coord_t col, FAR uint8_t *buffer,
+                         size_t npixels);
+
+/* LCD Configuration */
+
+static int apa102_getvideoinfo(FAR struct lcd_dev_s *dev,
+                               FAR struct fb_videoinfo_s *vinfo);
+static int apa102_getplaneinfo(FAR struct lcd_dev_s *dev,
+                               unsigned int planeno,
+                               FAR struct lcd_planeinfo_s *pinfo);
+
+/* LCD Specific Controls */
+
+static int apa102_getpower(FAR struct lcd_dev_s *dev);
+static int apa102_setpower(FAR struct lcd_dev_s *dev, int power);
+static int apa102_getcontrast(FAR struct lcd_dev_s *dev);
+static int apa102_setcontrast(FAR struct lcd_dev_s *dev,
+                               unsigned int contrast);
+
+/* Initialization */
+
+static inline void up_clear(FAR struct apa102_dev_s  *priv);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* This is working memory allocated by the LCD driver for each LCD device
+ * and for each color plane.  This memory will hold one raster line of data.
+ * The size of the allocated run buffer must therefore be at least
+ * (bpp * xres / 8).  Actual alignment of the buffer must conform to the
+ * bitwidth of the underlying pixel type.
+ *
+ * If there are multiple planes, they may share the same working buffer
+ * because different planes will not be operate on concurrently.  However,
+ * if there are multiple LCD devices, they must each have unique run buffers.
+ */
+
+static uint8_t g_runbuffer[APA102_LUT_SIZE];
+
+/* This structure describes the overall LCD video controller */
+
+static const struct fb_videoinfo_s g_videoinfo =
+{
+  APA102_COLORFMT,    /* Color format: RGB16-565: RRRR RGGG GGGB BBBB */
+  APA102_XRES,        /* Horizontal resolution in pixel columns */
+  APA102_YRES,        /* Vertical resolution in pixel rows */
+  1,                  /* Number of color planes supported */
+};
+
+/* This is the standard, NuttX Plane information object */
+
+static const struct lcd_planeinfo_s g_planeinfo =
+{
+  .putrun = apa102_putrun,              /* Put a run into LCD memory */
+  .getrun = apa102_getrun,              /* Get a run from LCD memory */
+  .buffer = (FAR uint8_t *)g_runbuffer, /* Run scratch buffer */
+  .bpp    = APA102_BPP,                 /* Bits-per-pixel */
```

Comment:
can we use C89 init?

Reply:
Ok, but it is funny because even skeleton.c is not using C89! Hehehe